server/src/main/java/org/opensearch/authz/WriteableParameter.java

```diff
@@ -0,0 +1,197 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.authz;
+
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.common.unit.ByteSizeValue;
+import org.opensearch.common.unit.TimeValue;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wrapper for {@link CheckableParameter} with {@link #writeTo(StreamOutput)} implementation dependent on the parameter type.
+ *
+ * @opensearch.experimental
+ */
+public class WriteableParameter implements Writeable {
+
+    /**
+     * The Generic Types which this class can serialize.
+     */
+    public enum ParameterType {
+        Boolean,
+        Integer,
+        Long,
+        Float,
+        Double,
+        String,
+        TimeValue, // long + TimeUnit
+        ByteSizeValue, // long + ByteSizeUnit
+    }
+
+    private CheckableParameter<?> parameter;
+    private ParameterType type;
+
+    /**
+     * Wrap a {@link CheckableParameter}. The generic type is determined from the type of the value.
+     *
+     * @param parameter  The parameter to wrap. The value must not be null.
+     * @throws IllegalArgumentException if the parameter has a null value.
+     */
+    public WriteableParameter(CheckableParameter<?> parameter) {
+        this(parameter, ParameterType.valueOf(parameter.getType().getSimpleName()));
+    }
+
+    /**
+     * Wrap a {@link CheckableParameter} with a specified generic type.
+     *
+     * @param parameter  The parameter to wrap.
+     * @param type  The Generic type of the parameter.
+     */
+    public WriteableParameter(CheckableParameter<?> parameter, ParameterType type) {
+        this.parameter = parameter;
+        this.type = type;
+    }
+
+    /**
+     * Wrap a {@link CheckableParameter} read from a stream.
+     *
+     * @param in Input to read the value from.
+     * @throws IOException if there is an error reading the values
+     */
+    public WriteableParameter(StreamInput in) throws IOException {
+        // Read the type
+        this.type = in.readEnum(ParameterType.class);
+        // Read the key
+        String key = in.readString();
+        // Read the value
+        Object value = readValue(in);
+        this.parameter = createParameter(type, key, value);
+    }
+
+    /**
+     * Gets the wrapped setting. Use {@link #getType()} to determine its generic type.
+     *
+     * @return The wrapped setting.
+     */
+    public CheckableParameter<?> getParameter() {
+        return this.parameter;
+    }
+
+    /**
+     * Gets the generic type of the wrapped setting.
+     *
+     * @return The wrapped setting's generic type.
+     */
+    public ParameterType getType() {
+        return this.type;
+    }
+
+    @SuppressWarnings("unchecked")
+    private CheckableParameter<?> createParameter(ParameterType type, String key, Object value) {
```

Comment:
This doesn't make sure callers use the correct value with the correct ParameterType so they can get an ugly runtime exception.  Instead what about could create a set of overloads like `createParameter(String key, TimeValue value)` and `createParameter(String key, ByteSizeValue value)` that create the background implementations.

I'm uncertain the worth going back and forth on this, but I think we might want to revisit after we've implemented more classes